scrapy/spiders/crawl.py

```diff
@@ -74,48 +81,61 @@ def parse_start_url(self, response):
     def process_results(self, response, results):
         return results
 
-    def _build_request(self, rule, link):
-        r = Request(url=link.url, callback=self._response_downloaded)
-        r.meta.update(rule=rule, link_text=link.text)
-        return r
+    def _build_request(self, rule_index, link):
+        return Request(
+            url=link.url,
+            callback=self._callback,
+            errback=self._errback,
+            meta=dict(rule=rule_index, link_text=link.text),
+        )
 
     def _requests_to_follow(self, response):
         if not isinstance(response, HtmlResponse):
             return
         seen = set()
-        for n, rule in enumerate(self._rules):
-            links = [lnk for lnk in rule.link_extractor.extract_links(response)
-                     if lnk not in seen]
-            if links and rule.process_links:
-                links = rule.process_links(links)
-            for link in links:
+        for rule_index, rule in enumerate(self._rules):
+            links = (lnk for lnk in rule.link_extractor.extract_links(response)
+                     if lnk not in seen)
+            for link in rule.process_links(links):
                 seen.add(link)
-                request = self._build_request(n, link)
+                request = self._build_request(rule_index, link)
                 yield rule._process_request(request, response)
 
-    def _response_downloaded(self, response):
+    def _callback(self, response):
         rule = self._rules[response.meta['rule']]
         return self._parse_response(response, rule.callback, rule.cb_kwargs, rule.follow)
 
+    def _errback(self, failure):
+        rule = self._rules[failure.request.meta['rule']]
+        return self._handle_failure(failure, rule.errback)
+
     def _parse_response(self, response, callback, cb_kwargs, follow=True):
         if callback:
             cb_res = callback(response, **cb_kwargs) or ()
             cb_res = self.process_results(response, cb_res)
-            for requests_or_item in iterate_spider_output(cb_res):
-                yield requests_or_item
+            for request_or_item in iterate_spider_output(cb_res):
+                yield request_or_item
 
         if follow and self._follow_links:
             for request_or_item in self._requests_to_follow(response):
                 yield request_or_item
 
+    def _handle_failure(self, failure, errback):
+        if errback:
+            results = errback(failure) or ()
+            if hasattr(failure.value, 'response'):
+                results = self.process_results(failure.value.response, results)
```

Comment:
For the record, I'm thinking a good change here would be to call `process_results` with the failure object, and document it as `process_results(response_or_failure, results)`, but since the method also exists on other specific spider I think it might be better keep the scope of this PR as is and make a new one for those changes.